### PR TITLE
chore: introduce load table total rows

### PR DESCRIPTION
**Pull request by achettyiitr**
# Description

- Removed `GetTotalCountInTable` from integrations package.
- Added `LoadTableStats` to calculate the **rows inserted and updated** while loading the table.
- Centralized `dedup_rows` instead of calling from all destinations.
- Avoid logging SQL queries in Info logs. Currently, we are polluting logs so much in the warehouse. If we still need Queries to be logged, we can set a `SlowQueryThreshold` to a minimal value.
- Added UNIT Test for LoadTable.

## Linear Ticket

- https://linear.app/rudderstack/issue/PIPE-209/investigate-why-total-count-does-not-honour-timeout-for-30-seconds
- https://linear.app/rudderstack/issue/PIPE-202/avoid-using-count-for-getting-total-count-where-ever-needed

## Security

- [x] The code changed/added as part of this pull request won't create any security issues with how the software is being used.

### CODE REVIEW

File: warehouse/integrations/snowflake/snowflake.go [hunk 1]
```diff
@@ -604,26 +627,52 @@ func (sf *Snowflake) copyInto(
...
+		count, err := strconv.Atoi(countString)
+		if err != nil {
+			return 0, fmt.Errorf("converting rows loaded: %w", err)
+		}
+		rowsInserted += int64(count)
+	}
+	if err := rows.Err(); err != nil {
+		return 0, fmt.Errorf("iterating over rows: %w", err)
+	}
+	return rowsInserted, nil
```
Comment (achettyiitr) [inline warehouse/integrations/snowflake/snowflake.go:675]: Didn't find a better way of doing this?

File: warehouse/types/types.go [hunk 2]
```diff
@@ -0,0 +1,6 @@
+package types
+
+type LoadTableStats struct {
```
[the remark this answers is not in the document]
Comment (achettyiitr) [reply warehouse/types/types.go:3]: Good point. Moving it to integrations package.

File: warehouse/integrations/azure-synapse/azure-synapse.go [hunk 3]
```diff
@@ -218,232 +222,394 @@ func (*AzureSynapse) IsEmpty(context.Context, model.Warehouse) (empty bool, err
...
-	for _, column := range previousColumnKeys {
-		if !slices.Contains(sortedColumnKeys, column) {
-			extraColumns = append(extraColumns, column)
-		}
-	}
 	fileNames, err := as.LoadFileDownLoader.Download(ctx, tableName)
-	defer misc.RemoveFilePaths(fileNames...)
+	defer func() {
+		misc.RemoveFilePaths(fileNames...)
+	}()
```
[the remark this answers is not in the document]
Comment (achettyiitr) [reply warehouse/integrations/azure-synapse/azure-synapse.go:245]: Thanks for bringing this up.

[the remark this answers is not in the document]
Comment (achettyiitr) [reply warehouse/integrations/azure-synapse/azure-synapse.go:245]: > Then looking at the internals it seems like Download is not cleaning up after itself. If there has been an error, all files that have been downloaded for nothing (and are not returned apparently) could be cleaned up by the downloader itself.

That is true, in case of an error it does not delete the files.

File: warehouse/integrations/azure-synapse/azure-synapse.go [hunk 4]
```diff
@@ -218,232 +222,394 @@ func (*AzureSynapse) IsEmpty(context.Context, model.Warehouse) (empty bool, err
...
+		return nil, "", fmt.Errorf("commit transaction: %w", err)
 	}
-	// deduplication process
+
+	log.Infow("completed loading")
+
+	return &types.LoadTableStats{
+		RowsInserted: rowsInserted - rowsDeleted,
+		RowsUpdated:  rowsDeleted,
+	}, stagingTableName, nil
```
[the remark this answers is not in the document]
Comment (achettyiitr) [reply warehouse/integrations/azure-synapse/azure-synapse.go:359]: Good point. Let me add a test for this.

File: warehouse/integrations/azure-synapse/azure-synapse.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (achettyiitr) [reply warehouse/integrations/azure-synapse/azure-synapse.go:245]: I will move the defer block after the error. No point in keeping up before error check.

[the remark this answers is not in the document]
Comment (achettyiitr) [reply warehouse/integrations/azure-synapse/azure-synapse.go:245]: Have a separate task and PR for this?

[the remark this answers is not in the document]
Comment (achettyiitr) [reply warehouse/integrations/azure-synapse/azure-synapse.go:245]: Tracking it [here](https://linear.app/rudderstack/issue/PIPE-286/integrations-cleanup-download-task).

File: warehouse/integrations/snowflake/snowflake.go [hunk 5]
```diff
@@ -603,63 +643,52 @@ func (sf *Snowflake) copyInto(
...
-		return fmt.Errorf("copy into table: %w", err)
+	columns, err := rows.Columns()
+	if err != nil {
+		return 0, fmt.Errorf("getting columns: %w", err)
+	}
+	_, index, found := lo.FindIndexOf(columns, func(item string) bool {
+		return item == "rows_loaded"
+	})
+	if !found {
+		return 0, nil
```
Comment (fracasula) [inline warehouse/integrations/snowflake/snowflake.go:664]: `if !found` we should at least get a warning, right? We expect it to be there always otherwise there is some kind of regression with the sdk, right? Also, a test here would be nice too.

Comment (achettyiitr) [reply warehouse/integrations/snowflake/snowflake.go:664]: Sure.

Comment (achettyiitr) [reply warehouse/integrations/snowflake/snowflake.go:664]: I am also exploring [COPY_HISTORY](https://docs.snowflake.com/en/sql-reference/functions/copy_history) for this as well.

Comment (achettyiitr) [reply warehouse/integrations/snowflake/snowflake.go:664]: Added a warn log.

Comment (achettyiitr) [reply warehouse/integrations/snowflake/snowflake.go:665]: We can't use [COPY_HISTORY](https://docs.snowflake.com/en/sql-reference/functions/copy_history) since parallel ingestion can also happen at the same time when Rudderstack is loading.

Comment (achettyiitr) [reply warehouse/integrations/snowflake/snowflake.go:665]: Going with the current approach.

File: warehouse/integrations/deltalake/deltalake_test.go [hunk 6]
```diff
@@ -379,6 +384,649 @@ func TestIntegration(t *testing.T) {
 			})
 		}
 	})
+
+	t.Run("Load Table", func(t *testing.T) {
```
Comment (achettyiitr) [inline warehouse/integrations/deltalake/deltalake_test.go:388]: Keeping this as part of Integration tests instead of Unit Tests, since starting for the cluster/SQL warehouse takes around 10-15 mins.

File: warehouse/integrations/clickhouse/clickhouse.go [hunk 7]
```diff
@@ -1032,9 +1034,25 @@ func (ch *Clickhouse) LoadUserTables(ctx context.Context) (errorMap map[string]e
...
+
+	err = ch.loadTable(ctx, tableName, ch.Uploader.GetTableSchemaInUpload(tableName))
+	if err != nil {
+		return nil, fmt.Errorf("loading table: %w", err)
+	}
+
+	postLoadTableCount, err := ch.totalCountIntable(ctx, tableName)
+	if err != nil {
+		return nil, fmt.Errorf("post load table count: %w", err)
+	}
```
Comment (achettyiitr) [inline warehouse/integrations/clickhouse/clickhouse.go:1051]: We can probably have flags later to disable it in case it's taking too long to execute. 
Currently, checked for existing customers and don't see any timeout error logs corresponding to `total count in table`.

File: warehouse/integrations/bigquery/bigquery.go [hunk 8]
```diff
@@ -793,41 +915,36 @@ func (bq *BigQuery) dropDanglingStagingTables(ctx context.Context) bool {
...
-		}
-		if !exists {
+		if exists, err := bq.tableExists(ctx, tableName); err != nil {
+			return false, fmt.Errorf("checking if table %s exists: %v", tableName, err)
+		} else if !exists {
 			continue
 		}
-		count, err := bq.GetTotalCountInTable(ctx, tableName)
+
+		metadata, err := bq.db.Dataset(bq.namespace).Table(tableName).Metadata(ctx)
```
Comment (fracasula) [inline warehouse/integrations/bigquery/bigquery.go:941]: This line doesn't seem to have coverage (many others don't but I think it would make sense to have it here). wdyt?

File: warehouse/integrations/bigquery/bigquery.go [hunk 9]
```diff
@@ -793,41 +915,36 @@ func (bq *BigQuery) dropDanglingStagingTables(ctx context.Context) bool {
...
 		if err != nil {
-			return empty, err
-		}
-		if count > 0 {
-			empty = false
-			return empty, nil
+			return false, fmt.Errorf("getting metadata for table %s: %v", tableName, err)
 		}
+		return metadata.NumRows == 0, nil
```
Comment (fracasula) [inline warehouse/integrations/bigquery/bigquery.go:945]: This is also not tested. I suggest you have a look at the codecov reports yourself and double check to verify that at least the main paths are tested.

File: warehouse/integrations/bigquery/bigquery.go [hunk 10]
```diff
@@ -793,41 +915,36 @@ func (bq *BigQuery) dropDanglingStagingTables(ctx context.Context) bool {
...
 	}
-	return
+	return true, nil
```
Comment (fracasula) [inline warehouse/integrations/bigquery/bigquery.go:947]: Also not under test. I guess we end up here if none of the `[]tables` exist.

File: warehouse/integrations/bigquery/testhelper/setup.go [hunk 11]
```diff
@@ -34,3 +44,35 @@ func IsBQTestCredentialsAvailable() bool {
...
+	t.Helper()
+
+	it, err := db.Query(query).Read(context.Background())
+	require.NoError(t, err)
+
+	var records [][]string
+	for {
+		var row []bigquery.Value
+		err = it.Next(&row)
+		if err == iterator.Done {
```
Comment (fracasula) [inline warehouse/integrations/bigquery/testhelper/setup.go:62]: I think by default it would make sense to avoid error comparisons in favour of `errors.Is` even when using a client directly. It would save us the trouble of forgetting to refactor this in case we were to use a wrapper or a different client that wraps the error. wdyt?

File: warehouse/integrations/bigquery/testhelper/setup.go [hunk 12]
```diff
@@ -34,3 +44,35 @@ func IsBQTestCredentialsAvailable() bool {
 	_, err := GetBQTestCredentials()
 	return err == nil
 }
+
+func RecordsFromWarehouse(
```
Comment (fracasula) [inline warehouse/integrations/bigquery/testhelper/setup.go:48]: What is this function actually doing? It's hard to tell without reading its code due to its generic name and lack of comment/docs. We should be able to make it more descriptive.

File: warehouse/integrations/mssql/mssql.go [hunk 13]
```diff
@@ -255,227 +253,344 @@ func (ms *MSSQL) DeleteBy(ctx context.Context, tableNames []string, params wareh
...
+			value = value[:mssqlStringLengthLimit]
+		}
+		if !hasDiacritics(value) {
+			return value, nil
+		} else {
+			byteArr := str2ucs2(value)
+			if len(byteArr) > mssqlStringLengthLimit {
+				byteArr = byteArr[:mssqlStringLengthLimit]
+			}
+			return byteArr, nil
```
Comment (fracasula) [inline warehouse/integrations/mssql/mssql.go:497]: This bit is also not tested.

File: warehouse/integrations/postgres/load.go [hunk 14]
```diff
@@ -18,172 +21,231 @@ import (
...
+		logfield.LoadTableStrategy, pg.loadTableStrategy(),
 	)
+	log.Infow("started loading")
 
-	return nil
-}
+	log.Infow("setting search path")
+	searchPathStmt := fmt.Sprintf(`
+		SET search_path TO %q;
+`,
```
Comment (fracasula) [inline warehouse/integrations/postgres/load.go:88]: [optional] Can't you format it without inserting more bytes into the query that we'll send over with the SQL client?

```suggestion
	searchPathStmt := fmt.Sprintf(`SET search_path TO %q;`,
```

File: warehouse/integrations/postgres/testdata/docker-compose.postgres.yml [hunk 15]
```diff
@@ -0,0 +1,15 @@
+version: "3.9"
+
+services:
+  postgres:
+    image: postgres:15-alpine
```
Comment (fracasula) [inline warehouse/integrations/postgres/testdata/docker-compose.postgres.yml:5]: @lvrach shouldn't we mention in [our docs](https://www.rudderstack.com/docs/destinations/warehouse-destinations/postgresql/) a minimum version that we're compatible with? :thinking:

File: warehouse/integrations/testhelper/setup.go [hunk 16]
```diff
@@ -242,3 +250,166 @@ func EnhanceWithDefaultEnvs(t testing.TB) {
...
+	uploadOutput, err := fm.Upload(
+		context.Background(), f, loadObjectFolder,
+		tableName, sourceID, uuid.New().String()+"-"+tableName,
+	)
+	require.NoError(t, err)
+
+	return uploadOutput
+}
+
+func RecordsFromWarehouse(
```
Comment (fracasula) [inline warehouse/integrations/testhelper/setup.go:278]: Is this function the same as the other helper?

File: warehouse/integrations/testhelper/setup.go [hunk 17]
```diff
@@ -242,3 +250,166 @@ func EnhanceWithDefaultEnvs(t testing.TB) {
...
+				return item.Format(time.RFC3339)
+			default:
+				return cast.ToString(item)
+			}
+		}))
+	}
+	return records
+}
+
+func LoadRecords() [][]string {
```
Comment (fracasula) [inline warehouse/integrations/testhelper/setup.go:316]: Is this loading any records though? It's just returning a matrix of slices.

Comment (fracasula) [reply warehouse/integrations/testhelper/setup.go:316]: Same comment goes for the other functions.

File: warehouse/integrations/snowflake/snowflake.go [hunk 18]
```diff
@@ -604,62 +632,58 @@ func (sf *Snowflake) copyInto(
...
 
-	return nil
-}
+	_, index, found := lo.FindIndexOf(columns, func(item string) bool {
+		return strings.ToLower(item) == "rows_loaded"
+	})
+	if !found {
+		sf.logger.Warnw("rows_loaded column not found in copy command result", "columns", columns)
+		return &types.LoadTableStats{}, nil
+	}
```
Comment (fracasula) [inline warehouse/integrations/snowflake/snowflake.go:654]: Not under test.

File: warehouse/integrations/bigquery/bigquery.go (hunk 9 above)
Comment (achettyiitr) [reply warehouse/integrations/bigquery/bigquery.go:945]: Thanks for pointing this out. I added test coverage considering only Load Table in mind.

Comment (achettyiitr) [reply warehouse/integrations/bigquery/bigquery.go:945]: let me re-verify and add coverage.

File: warehouse/integrations/bigquery/testhelper/setup.go (hunk 11 above)
Comment (achettyiitr) [reply warehouse/integrations/bigquery/testhelper/setup.go:62]: Yes.

File: warehouse/integrations/bigquery/bigquery.go (hunk 10 above)
Comment (achettyiitr) [reply warehouse/integrations/bigquery/bigquery.go:947]: Yes.

File: warehouse/integrations/testhelper/setup.go (hunk 16 above)
Comment (achettyiitr) [reply warehouse/integrations/testhelper/setup.go:278]: Renamed to RetrieveRecordsFromWarehouse with the comment.

File: warehouse/integrations/testhelper/setup.go (hunk 17 above)
Comment (achettyiitr) [reply warehouse/integrations/testhelper/setup.go:316]: These are helpers for the data we loaded in different scenarios. Update the function name and the added comment as well.

**Comment by achettyiitr** [comment]
@fracasula Thanks for pointing out missing codecov scenarios.

**Comment by fracasula** [comment]
@achettyiitr you got some conflicts on `warehouse/schema.go`.

**Review by fracasula — APPROVED** [review]

File: warehouse/integrations/postgres/testdata/docker-compose.postgres.yml (hunk 15 above)
Comment (achettyiitr) [reply warehouse/integrations/postgres/testdata/docker-compose.postgres.yml:5]: Good point. Currently, I can't find anything in [docs](https://www.rudderstack.com/docs/resources/rudderstack-architecture/#database) or [Readme](https://github.com/rudderlabs/rudder-server/blob/master/README.md) mentioning it.

File: warehouse/upload.go [hunk 19]
```diff
@@ -1122,34 +1068,25 @@ func (job *UploadJob) loadTable(tName string) (bool, error) {
...
-		if errEventCount != nil {
-			return
-		}
+	tableUpload, errEventCount := job.tableUploadsRepo.GetByUploadIDAndTableName(job.ctx, job.upload.ID, tName)
+	if errEventCount != nil {
+		return alteredSchema, fmt.Errorf("get table upload: %w", errEventCount)
+	}
 
-		// TODO : Perform the comparison here in the codebase
-		job.guageStat(`pre_load_table_rows`, whutils.Tag{Name: "tableName", Value: strings.ToLower(tName)}).Gauge(int(totalBeforeLoad))
```
Comment (achettyiitr) [inline warehouse/upload.go:1149]: pre_load_table_rows we are not using in any way right now, in Grafana or Alerts.

File: warehouse/integrations/azure-synapse/azure-synapse.go [hunk 20]
```diff
@@ -195,255 +206,379 @@ func (as *AzureSynapse) connectionCredentials() *credentials {
...
-				} else {
-					recordInterface = append(recordInterface, value)
-				}
+	csvReader := csv.NewReader(gzipReader)
+
+	for {
+		record, err := csvReader.Read()
+		if err != nil {
+			if err == io.EOF {
+				break
```
Comment (lvrach) [inline warehouse/integrations/azure-synapse/azure-synapse.go:387]: [nit-pick]
```suggestion
		if err == EOF {
		   break
		} else if err == io.EOF {

```

File: warehouse/integrations/postgres/testdata/docker-compose.postgres.yml (hunk 15 above)
Comment (lvrach) [reply warehouse/integrations/postgres/testdata/docker-compose.postgres.yml:5]: what's the minimum version we support? 

For Postgres as rudder-server dependency, we already have a check https://github.com/rudderlabs/rudder-server/blob/a6f205db9f87fc02f5c5290ec30bfacd216f59bf/services/validators/envValidator.go#L193 it is 10 which make sense since is no longer support by postgres https://endoflife.date/postgresql

I believe we should have 11 as our oldest supported version and also check this during validation

**Review by lvrach — APPROVED** [review]

File: warehouse/integrations/postgres/testdata/docker-compose.postgres.yml (hunk 15 above)
Comment (achettyiitr) [reply warehouse/integrations/postgres/testdata/docker-compose.postgres.yml:5]: @lvrach Are you referring to destination validation here? During the validation for `Postgres` destination, we also check the version.

Comment (achettyiitr) [reply warehouse/integrations/postgres/testdata/docker-compose.postgres.yml:5]: We are currently not sure what min postgres we support as a destination.

File: warehouse/integrations/azure-synapse/azure-synapse.go (hunk 20 above)
Comment (achettyiitr) [reply warehouse/integrations/azure-synapse/azure-synapse.go:387]: Going with this:

```GO
		record, err := csvReader.Read()
		if err != nil {
			if errors.Is(err, io.EOF) {
				break
			}
			return fmt.Errorf("reading record from: %w", err)
		}
```

File: warehouse/integrations/postgres/testdata/docker-compose.postgres.yml (hunk 15 above)
Comment (fracasula) [reply warehouse/integrations/postgres/testdata/docker-compose.postgres.yml:5]: > We are currently not sure what min postgres we support as a destination.

:point_up: 

Also, might make sense to have our Warehouse integration test run twice, one for the minimum supported version and one for the latest.

File: warehouse/integrations/azure-synapse/azure-synapse.go [hunk 21]
```diff
@@ -218,232 +222,394 @@ func (*AzureSynapse) IsEmpty(context.Context, model.Warehouse) (empty bool, err
...
-		}
-		gzipReader.Close()
-		gzipFile.Close()
+	}
+	if _, err = stmt.ExecContext(ctx); err != nil {
+		return nil, "", fmt.Errorf("executing copyIn statement: %w", err)
 	}
 
-	_, err = stmt.ExecContext(ctx)
+	log.Infow("deleting from load table")
```
[the remark this answers is not in the document]
Comment (achettyiitr) [reply warehouse/integrations/azure-synapse/azure-synapse.go:325]: I feel we should have this since it's a critical stage in our loading process. Currently, making these as debug logs. If we feel that some are still polluting it, we can change those later. wdyt?

```
log.Debugw("creating staging table")
log.Debugw("creating prepared stmt for loading data")
log.Debugw("committing transaction")
log.Debugw("waiting for load/append/merge job to complete", "jobID", job.ID())
log.Debugw("job statistics")
log.Debugw("setting search path")
log.Debugw("generated manifest", "manifestLocation", manifestLocation)
```

File: warehouse/integrations/postgres/testdata/docker-compose.postgres.yml (hunk 15 above)
Comment (achettyiitr) [reply warehouse/integrations/postgres/testdata/docker-compose.postgres.yml:5]: Corresponding task. https://linear.app/rudderstack/issue/PIPE-378/check-minimum-version-during-warehouse-validations

Comment (achettyiitr) [reply warehouse/integrations/postgres/testdata/docker-compose.postgres.yml:5]: Also, should we restrict this during the loading process as well? Maybe we can have a preliminary check with the existing customers, If everybody passes, we can enforce it else we need some action there.
